enricher/fabric8/src/main/java/io/fabric8/maven/enricher/fabric8/AutoTLSEnricher.java

```diff
@@ -0,0 +1,180 @@
+package io.fabric8.maven.enricher.fabric8;
+
+import io.fabric8.kubernetes.api.builder.TypedVisitor;
+import io.fabric8.kubernetes.api.model.ContainerBuilder;
+import io.fabric8.kubernetes.api.model.KubernetesListBuilder;
+import io.fabric8.kubernetes.api.model.PodSpecBuilder;
+import io.fabric8.kubernetes.api.model.PodTemplateSpecBuilder;
+import io.fabric8.kubernetes.api.model.Volume;
+import io.fabric8.kubernetes.api.model.VolumeMount;
+import io.fabric8.maven.core.util.Configs;
+import io.fabric8.maven.enricher.api.BaseEnricher;
+import io.fabric8.maven.enricher.api.EnricherContext;
+import io.fabric8.maven.enricher.api.Kind;
+import org.json.JSONArray;
+import org.json.JSONObject;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Enriches declarations with auto-TLS annotations, required secrets reference,
+ * mounted volumes and PEM to keystore converter init container.
+ *
+ * This is opt-in so should not be added to default enrichers as it only works for
+ * OpenShift.
+ */
+public class AutoTLSEnricher extends BaseEnricher {
+    static final String ENRICHER_NAME = "fmp-autotls";
+    static final String AUTOTLS_ANNOTATION_KEY = "service.alpha.openshift.io/serving-cert-secret-name";
+
+    private String secretName;
+
+    enum Config implements Configs.Key {
+        tlsSecretName,
+
+        tlsSecretVolumeMountPoint  {{ d = "/var/run/secrets/fabric8.io/tls-pem"; }},
+
+        tlsSecretVolumeName        {{ d = "tls-pem"; }},
+
+        jksVolumeMountPoint        {{ d = "/var/run/secrets/fabric8.io/tls-jks"; }},
+
+        jksVolumeName              {{ d = "tls-jks"; }},
+
+        pemToJKSInitContainerImage {{ d = "jimmidyson/pemtokeystore:v0.1.0"; }},
+
+        pemToJKSInitContainerName  {{ d = "tls-jks-converter"; }},
+
+        keystoreFileName           {{ d = "keystore.jks"; }},
+
+        keystorePassword           {{ d = "changeit"; }},
+
+        keystoreCertAlias          {{ d = "server"; }};
+
+        public String def() { return d; } protected String d;
+    }
+
+    public AutoTLSEnricher(EnricherContext buildContext) {
+        super(buildContext, ENRICHER_NAME);
+
+        this.secretName = getConfig(Config.tlsSecretName, getProject().getArtifactId() + "-tls");
+    }
+
+    @Override
+    public Map<String, String> getAnnotations(Kind kind) {
+        if (!isOpenShiftMode() || !kind.isService()) {
+            return null;
+        }
+        return Collections.singletonMap(AUTOTLS_ANNOTATION_KEY, this.secretName);
+    }
+
+    @Override
+    public void addMissingResources(KubernetesListBuilder builder) {
+        if (!isOpenShiftMode()) {
```

Comment:
While trying to fix other stuff, stumbled over this base `isOpenShiftMode()` method which has been introduced lately.

@jimmidyson @jstrachan @rawlingsj : Just a heads up (and also it was used for adding docker env vars to pod envs when using s2i binary build): It won't work as you might expect, since we always create kubernetes.yml and openshift.yml, regardless of the platform mode (auto detected or with -Dfabric8.mode). When the platform mode is used in an enricher it will create different kubernetes.yml and openshift.yml depending on the mode, but it still creates both files. And you can perfectly apply an openshift.yml to OpenShift which has been created with a `mvn -Dfabric8.mode=kubernetes fabric8:resource` (or when its running in auto detect mode and not being connected to an active OpenShift context). 

This is also dangerous when you refer to these files included in dependencies as it now matters with which mode these dependencies has been build.

The solution is of course again described in #678 so that actually kubernetes and openshift objects are not converted into each other in an after thought but by calling each enricher twice, once for kubernetes and once for openshift, independent on the fabric8.mode in use (which however still is needed for deciding how to do the build).

Until then this might work or not and for sake of simplicity I even will take this call over to fix the VolumePermissionEnricher to not add init containers for OpenShift.